File: src/test/resources/org/bau/benchmarks/go/linkedList.go
// https://github.com/lemire/Code-used-on-Daniel-Lemire-s-blog/blob/master/2026/02/13/stress_gc.go

package main

import (
    "fmt"
    "time"
)

type Node struct {
    value int
    next  *Node
}

func createLinkedList(len int) *Node {
    head := &Node{value: 0}
    current := head
    for i := 1; i < len; i++ {
        current.next = &Node{value: i}
        current = current.next
    }
    return head
}

func main() {
    fmt.Println("Go")
    x := createLinkedList(50_000_000)
    count := 1_000_000
    len := 1000
    worst := int64(0)
    start := time.Now().UnixNano()
    last := start
    for i := 0; i < count; i++ {
        createLinkedList(len)
        now := time.Now().UnixNano()
        worst = max(worst, now - last)
        last = now
    }
    avg := (time.Now().UnixNano() - start) / int64(count)
    fmt.Printf("Max delay: %d ms, avg %d ms; dummy=%d\n", 
        worst / 1_000_000.0, avg / 1_000_000.0, x.value)
}

File: src/test/resources/org/bau/benchmarks/rust/linkedList.rs
// https://lemire.me/blog/2026/02/15/how-bad-can-python-stop-the-world-pauses-get/

use std::time::Instant;

struct Node {
    value: i32,
    next: Option<Box<Node>>,
}

fn create_linked_list(len: usize) -> Box<Node> {
    let mut head = Box::new(Node {
        value: 0,
        next: None,
    });
    let mut current = &mut head;
    for i in 1..len {
        current.next = Some(Box::new(Node {
            value: i as i32,
            next: None,
        }));
        current = current.next.as_mut().unwrap();
    }
    head
}

impl Drop for Node {
    fn drop(&mut self) {
        // Take the next node out so Rust doesn't recursively drop it.
        let mut current = self.next.take();
        while let Some(mut boxed) = current {
            // Detach the next pointer before boxed goes out of scope.
            current = boxed.next.take();
            // boxed is dropped here, but its `next` is already None,
            // so no recursive drop happens.
        }
    }
}

fn main() {
    println!("Starting stress test...");
    let x = create_linked_list(5000);
    let count: usize = 1_000_000;
    let len: usize = 1000;
    let mut worst: u64 = 0;
    let mut last = Instant::now();
    for _ in 0..count {
        create_linked_list(len);
        let now = Instant::now();
        let elapsed = (now - last).as_nanos() as u64;
        worst = std::cmp::max(worst, elapsed);
        last = now;
    }
    println!("Max delay: {} ms; dummy={}", (worst as f64) / 1_000_000.0, x.value);
}